Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,434 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+public typealias NIOHTTPClientUpgradeConfiguration = (upgraders: [NIOHTTPClientProtocolUpgrader], completionHandler: (ChannelHandlerContext) -> Void)
+
+/// Errors that may be raised by the `HTTPClientProtocolUpgrader`.
+public enum NIOHTTPClientUpgradeError: Error {
+    case responseProtocolNotFound
+    case invalidHTTPOrdering
+    case upgraderDeniedUpgrade
+}
+
+/// An object that implements `ProtocolUpgrader` knows how to handle HTTP upgrade to
+/// a specific protocol.
+public protocol NIOHTTPClientProtocolUpgrader {
+    
+    /// The protocol this upgrader knows how to support.
+    var supportedProtocol: String { get }
+    
+    /// All the header fields the protocol requires in the request to successfully upgrade.
+    /// These header fields will be added to the outbound request's "Connection" header field.
+    /// It is the responsibility of the custom headers call to actually add these required headers.
+    var requiredUpgradeHeaders: [String] { get }
+    
+    /// Additional headers to be added to the request, beyond the "Upgrade" and "Connection" headers.
+    func addCustom(upgradeRequestHeaders: inout HTTPHeaders)
+    
+    /// Gives the receiving upgrader the chance to deny the upgrade based on the upgrade HTTP response.
+    func shouldAllowUpgrade(upgradeResponse: HTTPResponseHead) -> Bool
```

Comment:
@tanner0101 Do you have any need to delay accepting a protocol upgrade, such that this API would need to return a `Future`? Having a `Future` in the API here is somewhat dangerous from a protocol management perspective, so if we can have this function be required to be synchronous that would be ideal.

Reply:
I think it's fine being synchronous. 👍

Reply:
Thanks. Won't change it then.